p2p/subscriber.go

```diff
@@ -197,20 +178,46 @@ func (s *Subscriber[H]) verifyMessage(
 	}
 
 	var verErr *header.VerifyError
-	err = s.verifier(ctx, hdr)
-	switch {
+	switch err := s.verifier(ctx, hdr); {
 	case errors.As(err, &verErr) && verErr.SoftFailure:
 		s.metrics.ignore(ctx)
 		return pubsub.ValidationIgnore
 	case err != nil:
 		s.metrics.reject(ctx)
 		return pubsub.ValidationReject
 	default:
+		// keep the valid header in the msg so Subscriptions can access it without
+		// additional unmarshalling
+		msg.ValidatorData = hdr
+		s.metrics.accept(ctx, len(msg.Data))
+		return pubsub.ValidationAccept
 	}
+}
 
-	// keep the valid header in the msg so Subscriptions can access it without
-	// additional unmarshalling
-	msg.ValidatorData = hdr
-	s.metrics.accept(ctx, len(msg.Data))
-	return pubsub.ValidationAccept
+func (s *Subscriber[H]) extractHeader(ctx context.Context, p peer.ID, msg *pubsub.Message,
+) (H, bool) {
+	if msg.ValidatorData != nil {
+		hdr, ok := msg.ValidatorData.(H)
+		if !ok {
+			panic(fmt.Sprintf("msg ValidatorData is of type %T", msg.ValidatorData))
+		}
+		return hdr, true
+	}
+
+	hdr := header.New[H]()
+	if err := hdr.UnmarshalBinary(msg.Data); err != nil {
+		log.Errorw("unmarshalling header",
+			"from", p.ShortString(),
+			"err", err)
+		s.metrics.reject(ctx)
+		return hdr, false
+	}
+	if err := hdr.Validate(); err != nil {
+		log.Errorw("invalid header",
+			"from", p.ShortString(),
```

Comment:
optional nit: would also avoid passing peer just for the log, could be logged outside

Reply:
We've 2 different log messages, didn't have a plan to get rid of them. 

I'm fine to do 1 logging in `if` with `extractHeader` instead. Agree?

Reply:
the downside there tho is that we may lose on error message granularity/precision
it becomes more general rather then specific I mean
thus optional

Reply:
(i wish GH had realtime chat)

Reply:
done